prow/gitattributes/pattern.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitattributes
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	negativePrefix = "!"
+	patternDirSep  = "/"
+	zeroToManyDirs = "**"
+)
+
+type pattern struct {
+	pattern []string
+	isGlob  bool
+}
+
+// parsePattern parses a gitattributes pattern string into the Pattern structure.
+// The rules by which the pattern matches paths are the same as in .gitignore files (see https://git-scm.com/docs/gitignore), with a few exceptions:
+//   - negative patterns are forbidden
+//   - patterns that match a directory do not recursively match paths inside that directory
+// https://git-scm.com/docs/gitattributes
+func parsePattern(p string) (Pattern, error) {
+	res := pattern{}
+
+	// negative patterns are forbidden
+	if strings.HasPrefix(p, negativePrefix) {
+		return nil, fmt.Errorf("negative patterns are forbidden: <%s>", p)
+	}
+
+	if !strings.HasSuffix(p, "\\ ") {
+		p = strings.TrimRight(p, " ")
+	}
+
+	// patterns that match a directory do not recursively match paths inside that directory
+	if strings.HasSuffix(p, patternDirSep) {
+		return nil, fmt.Errorf("directory patterns are not matched recursively, use path/** instead: <%s>", p)
+	}
+
+	if strings.Contains(p, patternDirSep) {
```

Comment:
Shouldn't this check for inclusion of the glob?

Reply:
Nope, I have renamed variables for clarity.